storage/local/doubledelta.go

```diff
@@ -0,0 +1,513 @@
+	if vb < d8 && isInt { // Only use int for fewer than 8 value double-delta bytes.
+		c[doubleDeltaHeaderIsIntOffset] = 1
+	} else {
+		c[doubleDeltaHeaderIsIntOffset] = 0
+	}
+	return &c
+}
+
+func (c doubleDeltaEncodedChunk) newFollowupChunk() chunk {
+	return newDoubleDeltaEncodedChunk(d1, d0, true, cap(c))
+}
+
+func (c doubleDeltaEncodedChunk) baseTime() clientmodel.Timestamp {
+	return clientmodel.Timestamp(
+		binary.LittleEndian.Uint64(
+			c[doubleDeltaHeaderBaseTimeOffset:],
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseValue() clientmodel.SampleValue {
+	return clientmodel.SampleValue(
+		math.Float64frombits(
+			binary.LittleEndian.Uint64(
+				c[doubleDeltaHeaderBaseValueOffset:],
+			),
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseTimeDelta() clientmodel.Timestamp {
+	return clientmodel.Timestamp(
+		binary.LittleEndian.Uint64(
+			c[doubleDeltaHeaderBaseTimeDeltaOffset:],
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) baseValueDelta() clientmodel.SampleValue {
+	return clientmodel.SampleValue(
+		math.Float64frombits(
+			binary.LittleEndian.Uint64(
+				c[doubleDeltaHeaderBaseValueDeltaOffset:],
+			),
+		),
+	)
+}
+
+func (c doubleDeltaEncodedChunk) timeBytes() deltaBytes {
+	return deltaBytes(c[doubleDeltaHeaderTimeBytesOffset])
+}
+
+func (c doubleDeltaEncodedChunk) valueBytes() deltaBytes {
+	return deltaBytes(c[doubleDeltaHeaderValueBytesOffset])
+}
+
+func (c doubleDeltaEncodedChunk) sampleSize() int {
+	return int(c.timeBytes() + c.valueBytes())
+}
+
+func (c doubleDeltaEncodedChunk) len() int {
+	if len(c) <= doubleDeltaHeaderIsIntOffset+1 {
+		return 0
+	}
+	if len(c) <= doubleDeltaHeaderBaseValueOffset+8 {
+		return 1
+	}
+	return (len(c)-doubleDeltaHeaderBytes)/c.sampleSize() + 2
+}
+
+func (c doubleDeltaEncodedChunk) isInt() bool {
+	return c[doubleDeltaHeaderIsIntOffset] == 1
+}
+
+// add implements chunk.
+func (c doubleDeltaEncodedChunk) add(s *metric.SamplePair) []chunk {
+	if len(c) <= doubleDeltaHeaderIsIntOffset+1 {
+		// This is the first sample added to this chunk. Add it as base
+		// time and value.
+		c = c[:doubleDeltaHeaderBaseValueOffset+8]
+		binary.LittleEndian.PutUint64(
+			c[doubleDeltaHeaderBaseTimeOffset:],
+			uint64(s.Timestamp),
+		)
+		binary.LittleEndian.PutUint64(
+			c[doubleDeltaHeaderBaseValueOffset:],
+			math.Float64bits(float64(s.Value)),
+		)
+		return []chunk{&c}
+	}
+
+	tb := c.timeBytes()
+	vb := c.valueBytes()
+
+	if len(c) <= doubleDeltaHeaderBaseValueOffset+8 {
+		// This is the 2nd sample added to this chunk. Calculate the
+		// base deltas.
+		baseTimeDelta := s.Timestamp - c.baseTime()
+		if baseTimeDelta < 0 {
+			// TODO(beorn7): We ignore this irregular case for now. Once
+			// https://github.com/prometheus/prometheus/issues/481 is
+			// fixed, we should panic here instead.
+			return []chunk{&c}
+		}
+		c = c[:doubleDeltaHeaderBytes]
+		if tb >= d8 || bytesNeededForUnsignedTimestampDelta(baseTimeDelta) >= d8 {
```

Comment:
The `>=` is a bit confusing for the reader here because it suggests that there is possibly something larger than d8. Maybe just `==`?